core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1967,39 +1967,138 @@ private[spark] object Utils extends Logging {
       conf: SparkConf,
       serviceName: String = ""): (T, Int) = {
```

Comment:
this method is now very large and difficult to follow. Is all the logic required? If so we should make smaller helper methods with good documentation.